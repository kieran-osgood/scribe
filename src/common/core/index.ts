// effect
export * as Boolean from '@effect/data/Boolean';
export * as Either from '@effect/data/Either';
export * from '@effect/data/Function';
export * as Option from '@effect/data/Option';
export * as Data from '@effect/data/Data';
// io
export * as Effect from '@effect/io/Effect';
export * as Runtime from '@effect/io/Runtime';
export * as Exit from '@effect/io/Exit';

//schema
export * as S from '@effect/schema/Schema';
export * as TF from '@effect/schema/TreeFormatter';

// export * as Sink from '@effect/stream/Sink';
// export * as Stream from '@effect/stream/Stream';

// core-ts
export * as IO from 'fp-ts/lib/IO.js';
export * as TE from 'fp-ts/lib/TaskEither.js';
export * as C from 'fp-ts/lib/Console.js';

// core-ts-std
export * as SStd from 'fp-ts-std/Struct';
