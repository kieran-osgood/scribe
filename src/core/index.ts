// effect
export * as Boolean from '@effect/data/Boolean';
export * as Chunk from '@effect/data/Chunk';
export * as Context from '@effect/data/Context';
export * as Data from '@effect/data/Data';
export { TaggedClass } from '@effect/data/Data';
export * as Either from '@effect/data/Either';
export * from '@effect/data/Function';
export * as Option from '@effect/data/Option';
export * as RA from '@effect/data/ReadonlyArray';
export * as R from '@effect/data/ReadonlyRecord';
export * as String from '@effect/data/String';
export * as T from '@effect/data/Tuple';
export * as Cause from '@effect/io/Cause';
export * as Layer from '@effect/io/Layer';
export * as Stream from '@effect/stream/Stream';

// io
export * as Effect from '@effect/io/Effect';
export * as Exit from '@effect/io/Exit';
export * as Runtime from '@effect/io/Runtime';

//schema
export * as S from '@effect/schema/Schema';
export * as TF from '@effect/schema/TreeFormatter';
