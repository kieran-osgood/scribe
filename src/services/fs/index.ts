// https://github.com/Effect-TS/node/pull/629/commits/dbbfc3b0f6f7e0b21bcca5bd095612ea998f49b6
export * from './error.js';
export * from './fs.js';
