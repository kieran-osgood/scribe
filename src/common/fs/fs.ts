import * as NFS from 'fs';
import path from 'path';
import { MkDirError, ReadFileError, StatError, WriteFileError } from './error';
import { Abortable } from 'node:events';
import { Context, Effect, pipe } from '@scribe/core';

export interface FS {
  writeFile: typeof NFS.writeFile;
  readFile: typeof NFS.readFile;
  mkdir: typeof NFS.mkdir;
  stat: typeof NFS.stat;
}

export const FS = Context.Tag<FS>();
export const FSLive = Context.make(FS, NFS);
export const writeFile = (
  file: NFS.PathOrFileDescriptor,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions
) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, WriteFileError, NFS.PathOrFileDescriptor>(resume => {
        fs.writeFile(file, data, options, error => {
          if (error) {
            resume(
              Effect.fail(new WriteFileError({ file, data, options, error }))
            );
          } else {
            resume(Effect.succeed(file));
          }
        });
      })
    )
  );

export const writeFileWithDir = (
  pathName: string,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions
): Effect.Effect<FS, WriteFileError | MkDirError, NFS.PathOrFileDescriptor> =>
  pipe(
    mkdir(path.dirname(pathName), { recursive: true }),
    Effect.flatMap(() => writeFile(pathName, data, options))
  );

export const readFile = (
  path: NFS.PathOrFileDescriptor,
  options:
    | ({
        encoding?: BufferEncoding;
        flag?: string | undefined;
      } & Abortable)
    | undefined
    | null
) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, ReadFileError, string | Buffer>(resume => {
        fs.readFile(path, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new ReadFileError({ path, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      })
    )
  );

export const mkdir = (
  file: NFS.PathLike,
  options: NFS.MakeDirectoryOptions & {
    recursive: true;
  }
) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, MkDirError, string | undefined>(resume => {
        fs.mkdir(file, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new MkDirError({ file, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      })
    )
  );

export const stat = (path: string) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, StatError, NFS.Stats>(resume =>
        fs.stat(path, (error, stats) => {
          if (error) {
            resume(Effect.fail(new StatError({ path, error })));
          } else {
            resume(Effect.succeed(stats));
          }
        })
      )
    )
  );

export const fileOrDirExists = (
  pathLike: string
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile() || _.isDirectory())
  );
