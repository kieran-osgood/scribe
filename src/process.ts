import { Context, Effect } from '@scribe/core';

export interface Process {
  cwd(): string;
}

export const Process = Context.Tag<Process>();

export const ProcessLive = Effect.provideService(Process, {
  cwd: process.cwd,
});

export const ProcessMock = Effect.provideService(Process, {
  cwd: () => '/mockdir',
});

export const createProcessMock = (cwd: string) =>
  Effect.provideService(Process, {
    cwd: () => cwd,
  });
