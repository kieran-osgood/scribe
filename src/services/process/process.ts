import { Context } from '@scribe/core';

export interface Process {
  cwd: () => string;
}

export const Process = Context.Tag<Process>();

export const ProcessLive = {
  cwd: () => process.cwd(),
} satisfies Process;

export const ProcessMock = {
  cwd: () => '/mockdir',
} satisfies Process;

export const createProcessMock = (cwd: string) => ({ cwd: () => cwd });

export const getProcess = (cwd: string) => {
  if (
    (process.env.NODE_ENV === 'production' ||
      process.env.NODE_ENV === 'development') &&
    !cwd
  ) {
    return ProcessLive;
  }

  // TODO: move this to top and allow passing in for users?
  if (cwd.length > 0) {
    return createProcessMock(cwd);
  }

  return ProcessMock;
};
