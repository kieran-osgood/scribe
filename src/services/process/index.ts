import { Context, Effect } from 'src/core';

export interface Index {
  cwd(): string;
}

export const Process = Context.Tag<Index>();

export const ProcessLive = Effect.provideService(Process, {
  cwd: process.cwd,
});

export const ProcessMock = Effect.provideService(Process, {
  cwd: () => '/mockdir',
});

export const createProcessMock = (cwd: string) =>
  Effect.provideService(Process, {
    cwd: () => cwd,
  });
