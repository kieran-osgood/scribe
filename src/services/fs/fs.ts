import { Abortable } from 'node:events';

import { Context, Effect, pipe } from 'effect';
import * as NFS from 'fs';
import * as memfs from 'memfs';
import path from 'path';

import { MkDirError, ReadFileError, StatError, WriteFileError } from './error';

export interface FS {
  writeFile: typeof NFS.writeFile;
  readFile: typeof NFS.readFile;
  mkdir: typeof NFS.mkdir;
  stat: typeof NFS.stat;
  open: typeof NFS.open;
  close: typeof NFS.close;
}

export const FS = Context.Tag<FS>();
export const FSLive = NFS;
export const FSMock = memfs.fs as unknown as typeof NFS;

export const getFS = (test: boolean) => {
  return test ? FSMock : FSLive;
};

export const writeFile = (
  file: NFS.PathOrFileDescriptor,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions,
) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, WriteFileError, NFS.PathOrFileDescriptor>(resume => {
        fs.writeFile(file, data, options, error => {
          if (error) {
            resume(
              Effect.fail(new WriteFileError({ file, data, options, error })),
            );
          } else {
            resume(Effect.succeed(file));
          }
        });
      }),
    ),
  );

export const writeFileWithDir = (
  pathName: string,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions,
): Effect.Effect<FS, WriteFileError | MkDirError, NFS.PathOrFileDescriptor> =>
  pipe(
    mkdir(path.dirname(pathName), { recursive: true }),
    Effect.flatMap(() => writeFile(pathName, data, options)),
  );

export const readFile = (
  path: NFS.PathOrFileDescriptor,
  options:
    | ({ encoding?: BufferEncoding; flag?: string } & Abortable)
    | undefined
    | null,
): Effect.Effect<FS, ReadFileError, string | Buffer> =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, ReadFileError, string | Buffer>(resume => {
        fs.readFile(path, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new ReadFileError({ path, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      }),
    ),
  );

export const mkdir = (
  file: NFS.PathLike,
  options: NFS.MakeDirectoryOptions & {
    recursive: true;
  } = {
    // TODO!!!! Check mode
    mode: 0,
    recursive: true,
  },
): Effect.Effect<FS, MkDirError, string | undefined> =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, MkDirError, string | undefined>(resume => {
        fs.mkdir(file, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new MkDirError({ file, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      }),
    ),
  );

export const stat = (path: string) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, StatError, NFS.Stats>(resume =>
        { fs.stat(path, (error, stats) => {
          if (error) {
            resume(Effect.fail(new StatError({ path, error })));
          } else {
            resume(Effect.succeed(stats));
          }
        }); },
      ),
    ),
  );

export const isFileOrDirectory = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile() || _.isDirectory()),
  );

export const isFile = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile()),
  );

export const isDirectory = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isDirectory()),
  );

// export const open = (
//   path: PathLike,
//   flags: OpenMode | undefined,
//   mode: Mode | undefined | null
// ): Effect.Effect<FS, OpenError, number> =>
//   pipe(
//     FS,
//     Effect.flatMap(fs =>
//       Effect.async<FS, OpenError, number>(resume =>
//         fs.open(path, flags, mode, (error, fd) => {
//           if (error) {
//             resume(Effect.fail(new OpenError({ path, flags, mode, error })));
//           } else {
//             resume(Effect.succeed(fd));
//           }
//         })
//       )
//     )
//   );

// export const close = (fd: number): Effect.Effect<FS, CloseError, number> =>
//   pipe(
//     FS,
//     Effect.flatMap(fs =>
//       Effect.async<FS, CloseError, number>(resume =>
//         fs.close(fd, error => {
//           if (error) {
//             resume(Effect.fail(new CloseError({ error, fd })));
//           } else {
//             resume(Effect.succeed(fd));
//           }
//         })
//       )
//     )
//   );
