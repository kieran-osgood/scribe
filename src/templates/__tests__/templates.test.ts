import { Effect, pipe, RA } from '@scribe/core';
import {
  constructTemplate,
  ConstructTemplateCtx,
  Ctx,
  writeTemplate,
  WriteTemplateCtx,
} from '../index';
import * as FS from '@scribe/fs';
import * as memfs from 'memfs';
import { vol } from 'memfs';
import path from 'path';
import * as Process from '../../process';

beforeEach(() => {
  vi.restoreAllMocks();
});

const screenFileContents = `describe('{{Name}}', function() {
  it('should ', function() {

  });
});`;

beforeEach(() => {
  vol.mkdirSync(process.cwd(), { recursive: true });
  vol.mkdirSync(path.join(process.cwd(), './test-fixtures'), {
    recursive: true,
  });
});
afterEach(() => vol.reset());

const mockConfig = {
  options: {
    templatesDirectories: ['test-fixtures'],
    rootOutDir: '',
  },
  templates: {
    screen: {
      outputs: [
        {
          templateFileKey: 'screen',
          output: {
            directory: 'test-fixtures/config',
            fileName: '{{Name}}.ts',
          },
        },
        {
          templateFileKey: 'screen.test',
          output: {
            directory: 'test-fixtures/config',
            fileName: '{{Name}}.test.ts',
          },
        },
      ],
    },
  },
};

const fileContents = 'TEST';

const _ctx = {
  config: mockConfig,
  input: {
    template: 'screen',
    name: 'login',
  },
  templateKeys: ['screen'],
} satisfies Ctx;

const templateOutput = {
  templateFileKey: 'screen',
  output: {
    fileName: '{{Name}}.ts', // good-scribe
    directory: 'test-fixtures/config',
  },
};

describe('writeTemplate', () => {
  it('should write file', () =>
    pipe(
      Effect.gen(function* ($) {
        const ctx = {
          fileContents,
          templateOutput,
          ..._ctx,
        } satisfies WriteTemplateCtx;
        const res = writeTemplate(ctx);
        const result = yield* $(res);
        expect(result).toBe(
          path.join(process.cwd(), '/test-fixtures/config/login.ts'),
        );

        const readResult = yield* $(
          FS.readFile('test-fixtures/config/login.ts', null),
        );
        expect(String(readResult)).toBe(fileContents);
      }),
      FS.FSMock,
      Process.ProcessLive,
      Effect.runPromise,
    ));
});

describe('constructTemplate', () => {
  it('should return fileContents formatted with variables', () =>
    pipe(
      Effect.gen(function* ($) {
        const ctx = {
          templateOutput: {
            templateFileKey: 'screen',
            output: {
              fileName: '{{Name}}.ts', // good-scribe
              directory: '',
            },
          },
          ..._ctx,
        } satisfies ConstructTemplateCtx;

        yield* $(
          FS.writeFileWithDir(
            path.join(process.cwd(), './test-fixtures/screen.scribe'),
            screenFileContents,
            null,
          ),
        );

        const result = yield* $(
          constructTemplate(ctx),
          Effect.map(RA.map(_ => _.fileContents)),
        );

        expect(result).toMatchInlineSnapshot(`
          [
            "describe('login', function() {
            it('should ', function() {

            });
          });",
          ]
        `);
      }),
      FS.FSMock,
      Process.ProcessLive,
      Effect.runPromise,
    ));

  it('should check process root dir for templates', () =>
    pipe(
      Effect.gen(function* ($) {
        const ctx = {
          ..._ctx,
          templateOutput: {
            templateFileKey: 'screen',
            output: {
              fileName: '{{Name}}.ts',
              directory: '',
            },
          },
          config: {
            templates: _ctx.config.templates,
          },
        } satisfies ConstructTemplateCtx;

        const rootDirScribePath = path.join(process.cwd(), '', 'screen.scribe');
        memfs.vol.writeFileSync(rootDirScribePath, screenFileContents);

        const result = yield* $(
          constructTemplate(ctx),
          Effect.map(RA.map(_ => _.fileContents)),
        );

        expect(result).toMatchInlineSnapshot(`
          [
            "describe('login', function() {
            it('should ', function() {

            });
          });",
          ]
        `);
      }),
      FS.FSMock,
      Process.ProcessLive,
      Effect.runPromise,
    ));

  it("should throw if scribe file isn't readable", () =>
    pipe(
      Effect.gen(function* ($) {
        const ctx = {
          templateOutput: {
            templateFileKey: 'BADKEY',
            output: {
              fileName: '', // good-scribe
              directory: '',
            },
          },
          ..._ctx,
        } satisfies ConstructTemplateCtx;

        const result = yield* $(constructTemplate(ctx), Effect.flip);

        expect(result).toBeInstanceOf(FS.ReadFileError);
      }),
      FS.FSMock,
      Process.ProcessMock,
      Effect.runPromise,
    ));
});
