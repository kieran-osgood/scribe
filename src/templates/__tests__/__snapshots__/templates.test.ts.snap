// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`constructTemplate > should return fileContents formatted with variables 1`] = `
{
  "_tag": "Chunk",
  "values": [
    "// @ts-ignore
import * as React from 'react';

type loginProps = {}
function loginScreen() {

}
",
  ],
}
`;
